File: 19.remove-nth-node-from-end-of-list.py
#
# @lc app=leetcode id=19 lang=python3
#
# [19] Remove Nth Node From End of List
#


# @lc code=start
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def removeNthFromEnd(self, head: Optional[ListNode], n: int) -> Optional[ListNode]:
        # count the length
        # the target node's idx: length-n
        length = 0
        curr = head
        while curr:
            length += 1
            curr = curr.next

        curr = head
        prev = None
        targetIdx = length - n
        currIdx = 0
        if targetIdx == 0:
            return head.next
        while currIdx != targetIdx:
            prev = curr
            curr = curr.next
            currIdx += 1

        prev.next = curr.next
        return head


# @lc code=end

File: 19.remove-nth-node-from-end-of-list.go
/*
 * @lc app=leetcode id=19 lang=golang
 *
 * [19] Remove Nth Node From End of List
 */

// @lc code=start
package main

/**
 * Definition for singly-linked list.
 * type ListNode struct {
 *     Val int
 *     Next *ListNode
 * }
 */
func removeNthFromEnd(head *ListNode, n int) *ListNode {
	total := 0
	curr := head
	for curr != nil {
		total++
		curr = curr.Next
	}
	idx := total - n // idx to remove
	if idx == 0 {
		return head.Next
	}
	i := 0
	var prev *ListNode
	curr = head
	for i <= idx {
		if i == idx {
			prev.Next = curr.Next
			curr.Next = nil
		}
		prev = curr
		curr = curr.Next
		i++
	}
	return head
}

// @lc code=end
